[llvm-readobj][COFF] Implement --coff-pseudoreloc in llvm-readobj to dump runtime pseudo-relocation records

kikairoya opened the pull request:
MinGW toolchain uses "runtime pseudo-relocation" mechanism to support auto-importing symbols from DLLs.
There is no commonly used tools for dump the pseudo-relocation records, so we implement that functionality in llvm-readobj.

kikairoya committed:
find pseudo-reloc symbols

kikairoya committed:
implement llvm-readobj --coff-pseudoreloc stub

kikairoya committed:
dump raw reloc data

kikairoya committed:
show symbol name

kikairoya committed:
use support::ulittle32_t

kikairoya committed:
use named functor instead of unnamed lambda

mstorsjo reviewed (commented):
Thanks for working on this, this is very much appreciated! The script has indeed been useful a number of times. Some of the times I needed it, I also considered writing something like this, but in that case, most of the cases where I needed it, I wanted to find pseudo relocations that were too narrow (32 bit pseudo relocations in 64 bit binaries), and wanted the more precise diagnostics about where it was caused (which I ask about in a review comment here as well). But in the end, at the time, I ended up implementing that warning in LLD instead, in 6daa4b90e28ef4789ae0eb3163c6770c110d3a28.

kikairoya commented:
Thank you for the review! I’ll work on addressing your comments and suggestions.

kikairoya committed:
dump target symbol

kikairoya committed:
note the option "Cygwin/MinGW specific" and reorder alphabetically

kikairoya committed:
fix handling of errors

- ignore properly
- use reportWarning instead of reportError
- uncapitalize message

kikairoya committed:
cache import directory size

kikairoya committed:
reorganize test

- use --implicit-check-not
- use obj2yaml
- use numeric capture pattern

kikairoya committed:
range check for reloc record region

kikairoya committed:
early exit when not a PE

kikairoya commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2000,6 +2001,217 @@ void COFFDumper::printCOFFBaseReloc() {
+void COFFDumper::printCOFFPseudoReloc() {
+  if (!Obj->getDOSHeader()) {
+    W.startLine()
+        << "pseudo-relocation is only meaningful for a PE image file\n";
```
I have a question - such this message should be a warning or a just printing?

jh7370 replied:
>  (I can't make the tests for these 2 conditions, because yaml2obj can't generate such corrupted binary. )

Optional, but you could (in a separate PR) look at enhancing yaml2obj to allow you to create such a corrupted binary. In ELF, we have several fields whose only real use-case is to allow us to test specific error cases. See for example tests that use fields like `SHName` and `SHOffset` in the ELF llvm-readobj tests.

kikairoya commented on llvm/tools/llvm-readobj/Opts.td:
```diff
@@ -81,12 +81,17 @@ def codeview : FF<"codeview", "Display CodeView debug information">, Group<grp_c
 def codeview_subsection_bytes : FF<"codeview-subsection-bytes", "Dump raw contents of codeview debug sections and records">, Group<grp_coff>;
-def coff_basereloc : FF<"coff-basereloc", "Display .reloc section">, Group<grp_coff>;
+def coff_basereloc : FF<"coff-basereloc", "Display .reloc section">,
+                     Group<grp_coff>;
```
`git clang-format` folded this line. Should it be ignored?

mstorsjo replied:
It's somewhat expected that the formatting tools can tweak surrounding lines a little bit - but in this case, it's kinda far away. So I'm ok with excluding this change (although the CI formatting checker might complain and want you to include it), but I'm also ok with including it.

kikairoya reviewed (commented):
I believe I've addressed the comments. Thanks!

kikairoya commented:
The CI failure seems related to https://github.com/llvm/llvm-project/pull/152299; I can't find 'Linking CXX executable bin/llc' from build log so perhaps is it reused across jobs?

mstorsjo commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2002,51 +2002,94 @@ void COFFDumper::printCOFFBaseReloc() {
   auto Count = Obj->getNumberOfSymbols();
   if (Count == 0) {
-    W.startLine() << "The symbol table has been stripped\n";
+    W.startLine() << "the symbol table has been stripped\n";
```
No strong opinion either way here, just consistency with whatever the tools does in other cases sounds good.

kikairoya committed:
reduce usage of auto

kikairoya committed:
reorganize warnings

kikairoya committed:
use sorted vector

kikairoya committed:
reorganize test 1/2

- fix position of line breaks
- name intermediate output binary
- remove a sprious and doubled empty lines

kikairoya committed:
reorganize test 2/2

- make parameter more symbolic a bit
- use parametrized yaml
  * sed while generating source rather than testing

kikairoya committed:
add missing tests

kikairoya committed:
warn about missing imported symbol

kikairoya reviewed (commented):
Thank you for the review again!
I've made changes to address your comments and suggestions. Thanks!

kikairoya committed:
format

jh7370 commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2140,38 +2162,43 @@ void COFFDumper::printCOFFPseudoReloc() {
       if (Ite == ImportDirectories.begin())
-        return nullptr;
+        return createStringError(
+            "the reference of the symbol points out of the import table");
```
I'm not a massive fan of the warnings in this function. I expect some of them could do with the address values as context. Also, grammatically, some improvements are needed, but I'm not sure on the best wording, because of my lack of specific PE/COFF knowledge.

@mstorsjo, could you provide suggestions for improvements to these warnings in this function, please?

jh7370 commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2180,31 +2207,45 @@ void COFFDumper::printCOFFPseudoReloc() {
     W.printHex("Symbol", Reloc.Symbol);
-    if (const auto *Sym = ImportedSymbols.find(Reloc.Symbol))
-      W.printString("SymbolName", *Sym);
+    if (Expected<StringRef> SymOrErr = ImportedSymbols.find(Reloc.Symbol))
```
Same comment as above re. braces.

jh7370 commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2180,31 +2207,45 @@ void COFFDumper::printCOFFPseudoReloc() {
 
     W.printHex("Target", Reloc.Target);
-    if (auto Ite = RVASymbolMap.upper_bound(Reloc.Target.value());
+    if (auto Ite = llvm::upper_bound(
```
Same comment as above re. braces.

jh7370 commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2000,6 +2001,122 @@ void COFFDumper::printCOFFBaseReloc() {
+       i < Count && (!RelocBegin.getRawPtr() || !RelocEnd.getRawPtr()); ++i) {
+    auto Sym = Obj->getSymbol(i);
+    if (Sym.takeError())
+      continue;
```
Not sure if you've realised this, but `reportWarning` only emits the first instance of each identical warning, so if the same warning were to be reported lots of times, only the first one will appear in the actual warning output.

jh7370 replied:
My apologies, I got a bit mixed up.

`ObjDumper` (from which `COFFDumper` derives) has a `reportUniqueWarning` method, which you should use instead of `reportWarning`. There was discussion quite some time ago that we'd replace `reportWarning` entirely with `reportUniqueWarning` usage within llvm-readobj and I thought that had happened, but apparently not.

jh7370 commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2002,51 +2002,94 @@ void COFFDumper::printCOFFBaseReloc() {
     auto Sym = Obj->getSymbol(i);
-    if (Sym.takeError())
+    if (!Sym) {
+      consumeError(Sym.takeError());
```
See my other comment about how `reportWarning` deduplicates identical warnings. If the warnings aren't actually identical, but are noisy, there's probably a case for only reporting the first time (per input file).

jh7370 reviewed (commented):
I've not done a full review this time around, but there are plenty of things to address nonetheless. Thanks for working on this!

kikairoya committed:
format, style, wording (nfc)

kikairoya reviewed (commented):
Thank you the reviewing a lot!
I've made a change for the wording etc, only contains NFC portions.
I'm working on addressing the remaining comments.

kikairoya committed:
use reportUniqueWarning

kikairoya committed:
include the section number into the warning about pointed section differs

kikairoya committed:
include the pointed address into the warning about address reversal

kikairoya committed:
include the referenced address into the warning about not pointing an import symbol

kikairoya committed:
include the address into the warning about address overrun

kikairoya committed:
cleanup

kikairoya commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2002,11 +2002,12 @@ void COFFDumper::printCOFFBaseReloc() {
-    W.startLine()
-        << "pseudo-relocation is only meaningful for a PE image file\n";
+  ListScope D(W, "PseudoReloc");
+  W.flush();
```
It's an artifact during modifying the tests. Removed.

kikairoya committed:
format

kikairoya commented:
I've updated to include the context (symbol values, section number, referenced address) into the warnings and to use `reportUniqueWarning`. Thanks!

mstorsjo commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2000,6 +2001,264 @@ void COFFDumper::printCOFFBaseReloc() {
+  }
+
+  if (const uint32_t Begin = RelocBegin.getValue(), End = RelocEnd.getValue();
+      Data.size() <= Begin || Data.size() <= End) {
```
Nit: I find the comparisons easier to understand if they are in the form `Begin >= Data.size()`. And I think having `End == Data.size()` is fine, so for that part we'd only want to check `End > Data.size()`

mstorsjo reviewed (commented):
This is ok from me from a technical point of view, only one minor nit.

kikairoya committed:
swap the operands in the comparison and allow END == end of section

jh7370 commented on llvm/tools/llvm-readobj/COFFDumper.cpp:
```diff
@@ -2002,52 +2002,125 @@ void COFFDumper::printCOFFBaseReloc() {
+      Begin >= Data.size() || End > Data.size()) {
+    reportUniqueWarning(
+        createStringError("the marker symbol of runtime pseudo-relocation "
+                          "points past the end of the section: 0x%x",
```
I don't think it's 100% obvious that this value is the actual offset or the section end offset. Perhaps both should be present?

jh7370 reviewed (approved):
One minor suggestion, otherwise LGTM from my point of view, but please wait for @mstorsjo to approve too, to check the response to my suggestion (I'm off work most of next week, but don't want to hold this up).

kikairoya committed:
include the section size into the warning about address overrun

mstorsjo reviewed (approved):
LGTM
